state_manager.go

```diff
@@ -0,0 +1,291 @@
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/facebookgo/stackerr"
+)
+
+type StateManager struct {
+	*sync.RWMutex
+	Log                    Logger `inject:""`
+	replicaSet             *ReplicaSet
+	baseAddrs              string
+	currentReplicaSetState *ReplicaSetState
+	syncTryChan            chan struct{}
+
+	proxyToReal map[string]string
+	realToProxy map[string]string
+	proxies     map[string]*Proxy
+}
+
+func NewStateManager(replicaSet *ReplicaSet) *StateManager {
+	manager := &StateManager{
+		RWMutex:     &sync.RWMutex{},
+		replicaSet:  replicaSet,
+		baseAddrs:   replicaSet.Addrs,
+		proxyToReal: make(map[string]string),
+		realToProxy: make(map[string]string),
+		proxies:     make(map[string]*Proxy),
+	}
+	return manager
+}
+
+func (manager *StateManager) Start() error {
+	manager.Log.Debug("starting manager")
+	manager.Lock()
+	defer manager.Unlock()
+	var err error
+	manager.currentReplicaSetState, err = manager.generateReplicaSetState()
+	if err != nil {
+		return err
+		return errors.New(fmt.Sprintf("error starting statemanager, replicaset in flux: %v", err))
+	}
+	healthyAddrs := manager.currentReplicaSetState.Addrs()
+
+	// Ensure we have at least one health address.
+	if len(healthyAddrs) == 0 {
+		return stackerr.Newf("no healthy primaries or secondaries: %s", manager.replicaSet.Addrs)
+	}
+
+	manager.addProxies(healthyAddrs...)
+
+	for _, proxy := range manager.proxies {
+		go manager.startProxy(proxy)
+	}
+	return nil
+}
+
+func (manager *StateManager) KeepSynchronized(syncChan chan struct{}) {
+	manager.syncTryChan = syncChan
+	for {
+		select {
+		case <-manager.syncTryChan:
+			manager.Synchronize()
+		}
+	}
+}
+
+// Get new state for a replica set, and synchronize internal state.
+func (manager *StateManager) Synchronize() {
+	t := manager.replicaSet.Stats.BumpTime("replica.manager.time")
+	defer t.End()
+
+	manager.RLock()
+	newState, err := manager.generateReplicaSetState()
+	if err != nil {
+		manager.replicaSet.Stats.BumpSum("replica.manager.failed_state_check", 1)
+		manager.Log.Errorf("all nodes possibly down?: %s", err)
+		manager.RUnlock()
+		return
+	}
+
+	comparison, err := manager.getComparison(manager.currentReplicaSetState.lastRS, newState.lastRS)
+	if err != nil {
+		manager.replicaSet.Stats.BumpSum("replica.manager.failed_comparison", 1)
+		manager.Log.Errorf("Manager failed comparison %s", err)
+		manager.RUnlock()
+		return
+	}
+	manager.RUnlock() // all reads done
+
+	lockedTime := manager.replicaSet.Stats.BumpTime("replica.manager.time.locked")
+	defer lockedTime.End()
+
+	manager.Lock()
+	defer manager.Unlock()
+	if err = manager.addRemoveProxies(comparison); err != nil {
+		manager.replicaSet.Stats.BumpSum("replica.manager.failed_proxy_update", 1)
+		manager.Log.Errorf("Manager failed proxy update %s", err)
+		return
+	}
+
+	manager.stopStartProxies(comparison)
+	manager.currentReplicaSetState = newState
+	//manager.baseAddrs = manager.currentReplicaSetState.Addrs()
```

Comment:
I think we do wanna do this. This statement is opposite of what I said few hours ago :)

When we are adding new host to mongo fleet, with this we don't need to actually update the mongo.yml immediately, we can do it at later stage since dvara will auto discover it, so long there is 1 healthy seed.